packages/runtime/container-runtime/src/summary/summaryGenerator.ts

```diff
@@ -268,15 +273,38 @@ export class SummaryGenerator {
 		);
 
 		let summaryData: SubmitSummaryResult | undefined;
+
+		// Called when summarization fails at any stage. Validates that the right params are passed when a particular
+		// stage failed. Logs Summarize_cancel event.
 		const fail = (
 			errorCode: keyof typeof summarizeErrors,
 			error?: any,
+			submitFailureResult?: ISubmitSummaryFailureResult,
+			nackFailureResult?: INackSummaryResult,
 			properties?: SummaryGeneratorTelemetry,
-			nackSummaryResult?: INackSummaryResult,
 		) => {
-			// UploadSummary may fail with 429 and retryAfter - respect that
-			// Summary Nack also can have retryAfter, it's parsed below and comes as a property.
-			const retryAfterSeconds = getRetryDelaySecondsFromError(error);
+			let retryAfterSeconds: number | undefined;
+			switch (errorCode) {
```

Comment:
In terms of making this change as safe as possible, I'm not sure adding this extra validation / enforcement is worth it.  Maybe better to do later as part of other bigger refactors where we're not also introducing new behavior - what do you think?

Reply:
Just realized the alternative I'm suggesting isn't obvious - replace this `switch` with something like:

`const retryAfterSeconds = submitFailureResult.retryAfterSeconds ?? nackFailureResult.retryAfterSeconds;`

That's more generic, in line with the old code that would peel it off the error.  I totally like the more precise approach better, but it's more risky.

Reply:
I reverted most of the changes here. Please take a look.